feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go

```diff
@@ -71,33 +72,95 @@ func (s *RelayRouter) UpdateSession(
 		}
 	}
-	resultChan := make(chan *protos.UpdateSessionResponse, len(reqMap))
+
+	// Each request in reqMap is segregated into two requests and
+	// relays each request to respective proxies (session_proxy and n7n40_proxy).
+	// Hence length of resultChan is len(reqMap) * 2
+
+	resultChan := make(chan *protos.UpdateSessionResponse, len(reqMap)*2)
+	proxyMap := make(map[gateway_registry.GwServiceType]*protos.UpdateSessionRequest)
 
 	// send a separate Update request for each unique PLMN ID
 	for plmnid, req := range reqMap {
-		go func(plmnid string, req *protos.UpdateSessionRequest) {
-			client, ctx, cancel, err := s.getSessionProxyClient(ctx, plmnid)
-			if err != nil {
-				glog.Errorf("failed connect to Session Proxy for PLMNID '%s': %v", plmnid, err)
-				resultChan <- genUpdateErrorResp(req)
-				return
+		for _, pu := range req.GetUpdates() {
+			if pu == nil {
+				continue
 			}
-			defer cancel()
-			resp, err := client.UpdateSession(ctx, req)
-			if err != nil {
-				glog.Errorf("failed Session Proxy Update for PLMNID '%s': %v", plmnid, err)
-				resultChan <- genUpdateErrorResp(req)
-				return
+			if pu.CommonContext.RatType == protos.RATType_TGPP_NR {
+				n7n40ProxyReq, ok := proxyMap[FegN7N40Proxy]
+				if !ok || n7n40ProxyReq == nil {
+					n7n40ProxyReq = &protos.UpdateSessionRequest{
+						Updates:       []*protos.CreditUsageUpdate{},
+						UsageMonitors: []*protos.UsageMonitoringUpdateRequest{}}
+					proxyMap[FegN7N40Proxy] = n7n40ProxyReq
+				}
+				n7n40ProxyReq.Updates = append(n7n40ProxyReq.Updates, pu)
+			} else {
+				sessionProxyReq, ok := proxyMap[FegSessionProxy]
+				if !ok || sessionProxyReq == nil {
+					sessionProxyReq = &protos.UpdateSessionRequest{
+						Updates:       []*protos.CreditUsageUpdate{},
+						UsageMonitors: []*protos.UsageMonitoringUpdateRequest{}}
+					proxyMap[FegSessionProxy] = sessionProxyReq
+				}
+				sessionProxyReq.Updates = append(sessionProxyReq.Updates, pu)
 			}
-			resultChan <- resp
-		}(plmnid, req)
+		}
+
+		for _, pm := range req.GetUsageMonitors() {
+			if pm == nil {
+				continue
+			}
+			if pm.RatType == protos.RATType_TGPP_NR {
+				n7n40ProxyReq, ok := proxyMap[FegN7N40Proxy]
+				if !ok || n7n40ProxyReq == nil {
+					n7n40ProxyReq = &protos.UpdateSessionRequest{
+						Updates:       []*protos.CreditUsageUpdate{},
+						UsageMonitors: []*protos.UsageMonitoringUpdateRequest{}}
+					proxyMap[FegN7N40Proxy] = n7n40ProxyReq
+				}
+				n7n40ProxyReq.UsageMonitors = append(n7n40ProxyReq.UsageMonitors, pm)
+			} else {
+				sessionProxyReq, ok := proxyMap[FegSessionProxy]
+				if !ok || sessionProxyReq == nil {
+					sessionProxyReq = &protos.UpdateSessionRequest{
+						Updates:       []*protos.CreditUsageUpdate{},
+						UsageMonitors: []*protos.UsageMonitoringUpdateRequest{}}
+					proxyMap[FegSessionProxy] = sessionProxyReq
+				}
+				sessionProxyReq.UsageMonitors = append(sessionProxyReq.UsageMonitors, pm)
+			}
+		}
```

Comment:
try to use a bi dimensional map

I haven't checked but somethin like this

Remember to encapsulate in functions if you can
```
reqMap := map[string]map[string]*protos.UpdateSessionRequest{}
```

Then

```
for plmnid, req := range reqMap {
	for _, u := range r.GetUpdates() {
		if u == nil {
			continue
		}
		plmnid := getPlmnId6(u.GetCommonContext().GetSid().GetId())
		req, ok := reqMap[plmnid]
		if !ok || req == nil {
			addReqMapEntry(req)
		} 
		switch {
			req.CommonContext.RatType == protos.4g:
				reqMap[plmnid][FegSessionProxy].Updates = append(reqMap[plmnid][FegSessionProxy].Updates, u)
			req.CommonContext.RatType == protos.RATType_TGPP_NR:
				reqMap[plmnid][FegN7N40Proxy].Updates = append(reqMap[plmnid][FegN7N40Proxy].Updates , u)
			default:
				return error

	
		}
	}
	

func addReqMapEntry(reqMap map[string]map[string]*protos.UpdateSessionRequest{}){
	reqMap[plmnid][FegSessionProxy]= &protos.UpdateSessionRequest{
						Updates:       []*protos.CreditUsageUpdate{},
						UsageMonitors: []*protos.UsageMonitoringUpdateRequest{} 		 				

	reqMap[plmnid][FegN7N40Proxy]=  &protos.UpdateSessionRequest{
						Updates:       []*protos.CreditUsageUpdate{},
						UsageMonitors: []*protos.UsageMonitoringUpdateRequest{}}
}





```